src/NUnitCore/core/Compatibility.cs

```diff
@@ -0,0 +1,298 @@
+                        sourceName = (string)Reflect.GetPropertyValue(attribute, "SourceName");
+                        if (!string.IsNullOrEmpty(sourceName))
+                        {
+                            Type sourceType = (Type)Reflect.GetPropertyValue(attribute, "SourceType");
+                            if (sourceType == null && parameterInfo != null)
+                                sourceType = parameterInfo.Member.ReflectedType;
+                            if (sourceType != null)
+                            {
+                                var members = sourceType.GetMember(sourceName, ALL_MEMBERS);
+                                if (members.Length > 0 && !IsStaticMember(members[0]))
+                                    Error(sourceType + "." + members[0].Name, "ValueSourceAttribute must reference a static member.");
+                            }
+                        }
+                        break;
+
+                    case "System.STAThreadAttribute":
+                    case "System.MTAThreadAttribute":
+                        Warning(location, attributeFullName + " has no effect in NUnit 3. Use ApartmentAttribute.");
+                        break;
+                }
+            }
+        }
+
+        #endregion
+
+        #region Helper Methods
+
+        private static bool IsStaticMember(MemberInfo member)
+        {
+            switch (member.MemberType)
+            {
+                case MemberTypes.Field:
+                    return ((FieldInfo)member).IsStatic;
+                case MemberTypes.Property:
+                    var getMethod = ((PropertyInfo)member).GetGetMethod();
+                    return getMethod != null && getMethod.IsStatic;
+                case MemberTypes.Method:
+                    return ((MethodInfo)member).IsStatic;
+                default:
+                    return false;
+            }
+        }
+
+        private static void Report(string level, string location, string message)
+        {
+            if (_writer != null)
+                _writer.WriteLine(string.Format(REPORT_FORMAT, level, location, message));
+        }
+
+        private static string LocationOf(MemberInfo member)
+        {
+            return member.ReflectedType + "." + member.Name;
+        }
+
+        #endregion
+
+        #region Nested Issue Class
+
+        public class Issue
+        {
+            public string Level { get; private set; }
+            public string Location { get; private set; }
+            public string Message { get; private set; }
+
+            public Issue(string level, string category, string message)
+            {
+                Level = level;
+                Location = category;
+                Message = message;
+            }
+
+            public static Issue FromRecord(string record)
+            {
+                int delim1 = record.IndexOf(':');
+                int delim2 = record.IndexOf(':', delim1 + 1);
```

Comment:
Runtime assertions in case either of these somehow ends up `-1`?

Reply:
I didn't bother only because there is only one line that creates the records, and it always puts in two colons. But in principal that would be better.

BTW I didn't use Split here because the message (last component) may contain a colon.

Reply:
Maybe a good reason to make this method internal?

Reply:
Yup!